### PR TITLE
Allow intersections to be used as valid types for template literal placeholders

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/microsoft/TypeScript/issues/54177#issuecomment-1538436654

### CODE REVIEW

**Comment by gabritto** [comment]
Couple of things:
- There should be a fourslash test asserting that `"dataDowncast"` and other literals are suggested, since that's what the original issue is about.
- We're trying to add an official way to support this pattern of suggesting specific string literals even when the type of something is `string` so we can at some point stop supporting the `string & {}` workaround, so I'm not entirely sure we want to add even more code to support this.

**Comment by Andarist** [comment]
> There should be a fourslash test asserting that "dataDowncast" and other literals are suggested, since that's what the original issue is about.

done

> We're trying to add an official way to support this pattern of suggesting specific string literals even when the type of something is string so we can at some point stop supporting the string & {} workaround, so I'm not entirely sure we want to add even more code to support this.

Reading through some of the recent design notes  ([here](https://github.com/microsoft/TypeScript/issues/55105))... it's still unclear if/when/how the team is going to pick up this. This PR though doesn't introduce a lot of extra support for those "special intersections" - it **fixes** assignability issues when they are involved. This is a clear bug from the assignability PoV:
```ts
function conversionTest(groupName: | "downcast" | "dataDowncast" | "editingDowncast" | `${string & {}}Downcast`) {}
conversionTest("testDowncast"); // error but should be OK
```

Some other cases are improved with it as well, like this one:
```ts
function foo(str: `${`a${string}` & `${string}a`}Test`) {}
foo("abaTest"); // error but should be OK
```

**Comment by gabritto** [comment]
@typescript-bot test this
@typescript-bot run DT
@typescript-bot user test this
@typescript-bot user test tsserver
@typescript-bot test top100
@typescript-bot test tsserver top100
@typescript-bot perf test this

**Comment by gabritto** [comment]
@typescript-bot test this
@typescript-bot run DT
@typescript-bot user test this
@typescript-bot user test tsserver
@typescript-bot test top100
@typescript-bot test tsserver top100
@typescript-bot perf test this

**Comment by gabritto** [comment]
@typescript-bot pack this

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -24312,12 +24320,16 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         if (source === target || target.flags & (TypeFlags.Any | TypeFlags.String)) {
             return true;
         }
+        if (target.flags & TypeFlags.Intersection) {
+            return every((target as IntersectionType).types, t => t === emptyTypeLiteralType || isValidTypeForTemplateLiteralPlaceholder(source, t));
```
Comment (gabritto) [inline src/compiler/checker.ts:24324]: Why do we need to have `t === emptyTypeLiteralType` here? Couldn't `source` not be assignable to `{}`?

**Comment by gabritto** [comment]
It looks like this PR caused an out of memory error on a repo: https://github.com/microsoft/TypeScript/pull/54188#issuecomment-1662958205

**Comment by Andarist** [comment]
@gabritto I'm not sure if this failure is related to this PR. I can repro it but I can repro it using he build from `main` as well.

**Comment by gabritto** [comment]
@typescript-bot test tsserver top100

File: src/compiler/checker.ts (hunk 1 above)
Comment (Andarist) [reply src/compiler/checker.ts:24324]: `emptyTypeLiteralType` isn't a valid type for a template literal placeholder. Without this, we run into problems with what this PR is fixing:
<details>
<summary>test failures</summary>

```diff
diff --git a/tests/baselines/reference/templateLiteralTypesPatterns.errors.txt b/tests/baselines/reference/templateLiteralTypesPatterns.errors.txt
index cd82afc839..7bd15a6f82 100644
--- a/tests/baselines/reference/templateLiteralTypesPatterns.errors.txt
+++ b/tests/baselines/reference/templateLiteralTypesPatterns.errors.txt
@@ -55,10 +55,12 @@ templateLiteralTypesPatterns.ts(129,9): error TS2345: Argument of type '"1.1e-10
 templateLiteralTypesPatterns.ts(140,1): error TS2322: Type '`a${string}`' is not assignable to type '`a${number}`'.
 templateLiteralTypesPatterns.ts(141,1): error TS2322: Type '"bno"' is not assignable to type '`a${any}`'.
 templateLiteralTypesPatterns.ts(160,7): error TS2322: Type '"anything"' is not assignable to type '`${number} ${number}`'.
+templateLiteralTypesPatterns.ts(205,16): error TS2345: Argument of type '"testDowncast"' is not assignable to parameter of type '`${string & {}}Downcast` | "downcast" | "dataDowncast" | "editingDowncast"'.
+templateLiteralTypesPatterns.ts(207,17): error TS2345: Argument of type '"testDowncast"' is not assignable to parameter of type '"downcast" | "dataDowncast" | "editingDowncast" | `${{} & string}Downcast`'.
 templateLiteralTypesPatterns.ts(211,5): error TS2345: Argument of type '"abcTest"' is not assignable to parameter of type '`${`a${string}` & `${string}a`}Test`'.
 
 
-==== templateLiteralTypesPatterns.ts (58 errors) ====
+==== templateLiteralTypesPatterns.ts (60 errors) ====
     type RequiresLeadingSlash = `/${string}`;
     
     // ok
@@ -378,8 +380,12 @@ templateLiteralTypesPatterns.ts(211,5): error TS2345: Argument of type '"abcTest
     // repro from https://github.com/microsoft/TypeScript/issues/54177#issuecomment-1538436654
     function conversionTest(groupName: | "downcast" | "dataDowncast" | "editingDowncast" | `${string & {}}Downcast`) {}
     conversionTest("testDowncast");
+                   ~~~~~~~~~~~~~~
+!!! error TS2345: Argument of type '"testDowncast"' is not assignable to parameter of type '`${string & {}}Downcast` | "downcast" | "dataDowncast" | "editingDowncast"'.
     function conversionTest2(groupName: | "downcast" | "dataDowncast" | "editingDowncast" | `${{} & string}Downcast`) {}
     conversionTest2("testDowncast");
+                    ~~~~~~~~~~~~~~
+!!! error TS2345: Argument of type '"testDowncast"' is not assignable to parameter of type '"downcast" | "dataDowncast" | "editingDowncast" | `${{} & string}Downcast`'.
     
     function foo(str: `${`a${string}` & `${string}a`}Test`) {}
     foo("abaTest"); // ok
```
</details>

This was really meant to be a "faster"/more concise version of:
```ts
        if (target.flags & TypeFlags.Intersection) {
            if (areIntersectedTypesAvoidingPrimitiveReduction((target as IntersectionType).types)) {
                const primitive = (target as IntersectionType).types[0] === emptyTypeLiteralType ? (target as IntersectionType).types[1] : (target as IntersectionType).types[0];
                return isValidTypeForTemplateLiteralPlaceholder(source, primitive);
            }
            return every((target as IntersectionType).types, t => isValidTypeForTemplateLiteralPlaceholder(source, t));
        }
```

> Couldn't source not be assignable to {}?

This is a good question. I've tried to end up with such a source and I couldn't figure out a way to hit this. The whole existing test suite also doesn't hit such a case. I won't swear that it's impossible though so maybe it's better to use this alternative version that I posted above?

**Comment by Andarist** [comment]
> @gabritto I'm not sure if this failure is related to this PR. I can repro it but I can repro it using the build from main as well.

I might have misinterpreted the results in the rush. I **can't** repro this with either build (from this PR and from the main). Now I'm interested if the CI job will be able to repro it again 😅

File: src/compiler/checker.ts (hunk 1 above)
Comment (gabritto) [reply src/compiler/checker.ts:24324]: Ok, I looked at the code some more and I think this is actually ok (but like you said, I can't prove it's impossible).

**Comment by gabritto** [comment]
Seems like CI is fine now? I think this looks good for 5.3.

**Review by gabritto — APPROVED** [review]

**Pull request merged**